edb/server/dbview/dbview.pyx

```diff
@@ -887,6 +920,79 @@ cdef class DatabaseConnectionView:
+        rows = await conn.sql_fetch(b'SELECT "edgedb"."_clear_query_cache"()')
+        rv = []
+        for row in rows:
+            query_req = rpc.CompilationRequest(
+                self.server.compilation_config_serializer
+            ).deserialize(row[0], "<unknown>")
+            rv.append(query_req)
+            self._db._eql_to_compiled.pop(query_req, None)
+        return rv
+
+    async def recompile_all(
+        self, conn, requests: typing.Iterable[rpc.CompilationRequest]
+    ):
+        compiler_pool = self.server.get_compiler_pool()
+        concurrency = max(1, compiler_pool.get_size_hint() - 1)
+        i = asyncio.Queue(maxsize=concurrency)
+        o = asyncio.Queue()
+
+        async def recompile_request():
+            while True:
+                query_req = await i.get()
+                if query_req is None:
+                    o.put_nowait((None, None))
+                    break
+                try:
+                    result = await compiler_pool.compile(
+                        self.dbname,
+                        self.get_user_schema_pickle(),
+                        self.get_global_schema_pickle(),
+                        self.reflection_cache,
+                        self.get_database_config(),
+                        self.get_compilation_system_config(),
+                        query_req.serialize(),
+                        "<unknown>",
+                        client_id=self.tenant.client_id,
+                    )
+                except Exception:
+                    # discard cache entry that cannot be recompiled
+                    o.put_nowait((query_req, None))
+                    pass
+                else:
+                    o.put_nowait((query_req, result[0]))
+
+        async def persist_cache_task():
+            if not debug.flags.func_cache:
+                # TODO(fantix): sync _query_cache in one implicit tx
+                await conn.sql_fetch(b'SELECT "edgedb"."_clear_query_cache"()')
+
+            count = concurrency
+            while count > 0:
+                query_req, query_unit_group = await o.get()
+                if query_req is None:
+                    count -= 1
+                elif query_unit_group is None:
+                    self._db._eql_to_compiled.pop(query_req)
+                else:
+                    await execute.persist_cache(
+                        conn, self, query_req, query_unit_group
+                    )
+                    self._db._eql_to_compiled[query_req] = (
+                        query_unit_group, self.schema_version
+                    )
+
+        async with asyncio.TaskGroup() as g:
+            for _ in range(concurrency):
+                g.create_task(recompile_request())
+            g.create_task(persist_cache_task())
+            for data in requests:
+                await i.put(data)
+            for _ in range(concurrency):
+                await i.put(None)
```

Comment:
The code would be cleaner if we spun off one task per request and used a semaphore to limit concurrency, right?

I guess that would use more memory, though?

Reply:
Not too much more memory, I suppose; all the requests are already in memory.

Reply:
Yeah, just I think a task is bigger than a queue entry